packages/compose/src/hooks/use-focus-outside/index.ts

```diff
@@ -72,112 +21,48 @@ export default function useFocusOutside(
-		}
-	}, [] );
-
-	/**
-	 * A callback triggered when a blur event occurs on the element the handler
-	 * is bound to.
-	 *
-	 * Calls the `onFocusOutside` callback in an immediate timeout if focus has
-	 * move outside the bound element and is still within the document.
-	 */
-	const queueBlurCheck: React.FocusEventHandler = useCallback( ( event ) => {
-		// React does not allow using an event reference asynchronously
-		// due to recycling behavior, except when explicitly persisted.
-		event.persist();
-
-		// Skip blur check if clicking button. See `normalizeButtonFocus`.
-		if ( preventBlurCheckRef.current ) {
-			return;
-		}
-
-		// The usage of this attribute should be avoided. The only use case
-		// would be when we load modals that are not React components and
-		// therefore don't exist in the React tree. An example is opening
-		// the Media Library modal from another dialog.
-		// This attribute should contain a selector of the related target
-		// we want to ignore, because we still need to trigger the blur event
-		// on all other cases.
-		const ignoreForRelatedTarget = event.target.getAttribute(
-			'data-unstable-ignore-focus-outside-for-relatedtarget'
-		);
+		// Catch if a popover is spawning another popover.
+		// Does our container element contain an element with a popover, and that popover is expanded?
+		// If so, we don't want to run the onFocusOutside callback because it's spawning another popover
+		// and will need focus returned to it.
 		if (
-			ignoreForRelatedTarget &&
-			event.relatedTarget?.closest( ignoreForRelatedTarget )
+			containerElement?.querySelector(
+				'[aria-haspopup="true"][aria-expanded="true"]'
+			)
 		) {
 			return;
 		}
 
-		blurCheckTimeoutIdRef.current = setTimeout( () => {
-			// If document is not focused then focus should remain
-			// inside the wrapped component and therefore we cancel
-			// this blur event thereby leaving focus in place.
-			// https://developer.mozilla.org/en-US/docs/Web/API/Document/hasFocus.
-			if ( ! document.hasFocus() ) {
-				event.preventDefault();
-				return;
-			}
+		// Use event.relatedTarget instead of activeElement since activeElement and target
+		// might be body during the first blur event. Also, we're concerned with where focus
+		// is moving to (inside or outside the container), not where it came from.
+		const relatedTarget = event.relatedTarget as HTMLElement;
 
-			if ( 'function' === typeof currentOnFocusOutsideRef.current ) {
-				currentOnFocusOutsideRef.current( event );
-			}
-		}, 0 );
+		// Focus is outside the container and we should run the onFocusOutside callback if:
+		// - No target element
+		// - No container element
+		// - Target element is not contained within the container
+		if (
+			! relatedTarget ||
+			! containerElement ||
+			! containerElement.contains( relatedTarget )
+		) {
+			// Wrapping in setTimeout to run the check after focus and rendering resolves.
+			setTimeout( () => {
```

Comment:
Are we not able to get rid of the `setTimeout` ? Sorry if this was explained somewhere, I didn't see it.

Reply:
I think running the callback on blur is the wrong approach. It should run after focus resolves, because we'll get into weird states where the onClose callback sets states before the onMouseUp event finishes. This means you can't click a button to open the popover, then click the same button to close it (the popover will reopen). I think the `trunk` approach of queueing and canceling timeouts is a really clever and correct approach to this. 

I opened https://github.com/WordPress/gutenberg/pull/71294/ on top of this branch to remove all the extra browser hacks that I don't think are necessary anymore. It would simplify the code if we can remove it all. I'm not 100% sure if we still support a browser that needs that code though.